### ISSUE TITLE
Anyways to add device locally w/o IoT Cloud subscription?

### ISSUE THREAD

**Issue by anar4732** [jasonacox/tinytuya#249]
So the IoT Platform seems like having no free plan for individual developers like us. (only 1-month trial)
https://www.tuya.com/vas/commodity/IOT_CORE_V2

I will control devices locally. but unable to add my new devices since my trial was expired.
Is there any way to add devices locally?

**Comment by jasonacox** [comment]
Hi @anar4732  - Unfortunately, the cloud is the easiest way to get the Tuya device local key.

I'm curious, during the setup do you recall if you selected an account type?

<img width="706" alt="image" src="https://user-images.githubusercontent.com/836718/212247256-9a42eeae-2dc4-451c-9ce7-3f6f317fe814.png">

Also, are you able to sign up for a new account?

**Comment by anar4732** [comment]
> Hi @anar4732 - Unfortunately, the cloud is the easiest way to get the Tuya device local key.
> 
> I'm curious, during the setup do you recall if you selected an account type?
> 
> <img alt="image" width="706" src="https://user-images.githubusercontent.com/836718/212247256-9a42eeae-2dc4-451c-9ce7-3f6f317fe814.png">
> 
> Also, are you able to sign up for a new account?

do I need to skip selecting the account type?

**Comment by jasonacox** [comment]
Yes, in my testing, I "skip this step" and my accounts did not expire.  I have that instruction in the [PDF](https://github.com/jasonacox/tinytuya/files/8145832/Tuya.IoT.API.Setup.pdf) but should add that detail to the main readme if it works for you too.

**Comment by tjbcatch22** [comment]
I used cloud-cutter (`https://github.com/tuya-cloudcutter/tuya-cloudcutter`) on my bulb and am able to run my bulb without a cloud subscription, however, I have reduced functionality on the bulb (detailed below).  

For the details, in case this might work for you:
I've only used the Smart Home app to initially connect the bulb to get the --- version (1.2.16 in my case, for example).  I wonder if this is even necessary for subsequent bulbs of the same manufacture date code or lot 
My bulb seems to be a generic of some kind: Schwaiger A19 60W Equivalent Color Selectable Dimmable LED Light Bulb, (see here for more info: `https://www.menards.com/main/p-1642874274122411.htm`).
I run cloud-cutter and use Tinytuya bulb.py and can get the light to turn off, on, some dimming, and some limited modes. ( mode-scene has a softer light, mode-colour and music will turn the bulb red)

The bulb does not do any colors, except for when its in music or color mod as mentioned above.   I have not yet found a way to make the buld anything other than red.

**Comment by jasonacox** [comment]
Thanks @tjbcatch22 !  The [vulnerability and exploit](https://rb9.nl/posts/2022-03-29-light-jailbreaking-exploiting-tuya-iot-devices/) is fascinating.  @uzlonewolf have you seen this?  The timeline indicates that Tuya began patching so I wonder how long this path will remain open for people to avoid the cloud to get their keys.

@tjbcatch22 It does seem odd that this would disable color from your bulb. Were you able to set color before [tuya-cloudcutter](https://github.com/tuya-cloudcutter/tuya-cloudcutter)?  

When you say `bulb.py` are you using the example [here](https://github.com/jasonacox/tinytuya/blob/master/examples/bulb.py)?  I would recommend turning on debug by adding the line:

```python
# debug mode
tinytuya.set_debug(True)
```

It may give some clues as to what is happening during the `d.set_colour(r, g, b)` calls.

**Comment by tjbcatch22** [comment]
I'll check out the debug.  Thank you for mentioning that.  Stay tuned.  (perhaps I should create a new issue with the results so as not to bury the findings in this post)

Yes - The exploit if fascinating.  I'm wondering if that is why the bulb are on sale for 50% off.

**Comment by uzlonewolf** [comment]
@jasonacox Yes I'm familiar with it, I just didn't think there would be much overlap with TinyTuya users as it completely eliminates Tuya's cloud and app.  If someone was going to do that I figured they'd just install Tasmota or OpenBK7231N/OpenBeken (https://github.com/openshwprojects/OpenBK7231T_App) and be done with it.

**Comment by jasonacox** [comment]
Good point.  Having said that, the pure OTA possibility with tuya-cloudcutter could be attractive to some not wanting to break open their devices. I confess, I'm behind on catching up on what is possible with Tasmota these days so that could be a path there too.  But your point is still valid. I'm sure there is a large portion of our community that, like me, have cloud integrations they are not keen to break.  And, as I mentioned above, I suspect there is likely a limited exploit window on this OTA hack.  New firmware updates and devices will be coming from Tuya.

In the context of this issue, I do wish we could convince Tuya to provide an easier path for our home IoT maker community to get local keys without using their elaborate commercial Tuya IoT Developer process.

**Comment by tjbcatch22** [comment]
I would be in the camp of struggling with Tasmota or OpenBK7231N/OpenBeken.  I'm not a programmer - just an enthusiastic hobbyist - who is not compiling code very often.   Seems like OpenBK7231N would have a binary that I can use cloudcutter to upload, but im stuck on either finding the binary or building it.
Tinytuna, for me, is ready to go.  Easy python scripts.

**Comment by uzlonewolf** [comment]
Prebuild binaries are available on the 'releases' page https://github.com/openshwprojects/OpenBK7231T_App/releases .  I can understand not wanting to mess with it though.

**Comment by tjbcatch22** [comment]
Thank you!  Those binaries are great.  Looks like I was not able to find those on my own.

**Comment by uzlonewolf** [comment]
If you don't mind using node.js instead of Python then it looks like you can also get local keys via the "pretend to be the Android app" method.  A slight tweak to https://github.com/rospogrigio/localtuya/issues/1188 gets you everything with just your app username and password, no developer account required.  Simply add `console.log(device);` right after the `console.log('group: "%s"\tdevice: ...` line and run it.

**Comment by jasonacox** [comment]
Whoa! Nice! Should we implement this for the wizard?

**Comment by uzlonewolf** [comment]
I'm not thrilled with the idea of distributing the key/secret from a hacked Android app with tinytuya, but having the framework there so someone can just paste it in on their own would be good.

**Comment by jasonacox** [comment]
Excellent point.  That appears to be the approach tuyapi/cloud took as well.

Are these API keys likely only associated with one device and/or app install?

**Comment by uzlonewolf** [comment]
No, they're tied to the app version, they'll only change if a new version of the app is released.

**Comment by blakadder** [comment]
How about this [blakadder/tuya-uncover](https://github.com/blakadder/tuya-uncover)?

**Comment by uzlonewolf** [comment]
Very nice @blakadder !

**Comment by anar4732** [comment]
> Very nice @blakadder !

can it potentially implement on tinyTuya?

**Comment by uzlonewolf** [comment]
> can it potentially implement on tinyTuya?

Like I mentioned above, I'm not thrilled with the idea of distributing the key/secret from a hacked Android app with tinytuya, so if you can convince @jasonacox then sure.  Alternately, I wouldn't mind submitting a PR to [blakadder/tuya-uncover](https://github.com/blakadder/tuya-uncover) to add an option to make it spit out a TinyTuya-compatible devices.json file.

**Comment by jasonacox** [comment]
Thanks @anar4732 for the suggestion.  However, I agree completely with @uzlonewolf. We are not going to bundle the extracted Android app secrets into the PyPI module. Those may work today but could be disabled in the future. And more importantly, storing potentially confidential keys could violate PyPI's [Terms of Use](https://pypi.org/policy/terms-of-use/). 

I'm not opposed to having @blakadder 's script produce a compatible devices.json and referencing that link in the project as a helpful option.

**Comment by uzlonewolf** [comment]
I now have a PR in with @blakadder to add support for tinytuya-compatible device.json files.
